src/metatrain/experimental/phace/modules/layers.py

```diff
@@ -0,0 +1,186 @@
+from typing import Dict, List
+
+import torch
+from metatensor.torch import Labels, TensorBlock, TensorMap
+
+
+class Linear(torch.nn.Module):
+
+    def __init__(self, n_feat_in, n_feat_out):
+        super().__init__()
+        self.linear_layer = torch.nn.Linear(n_feat_in, n_feat_out, bias=False)
+        self.linear_layer.weight.data.normal_(0.0, 1.0)
+        self.n_feat_in = n_feat_in
+
+    def forward(self, x):
+        return self.linear_layer(x) * self.n_feat_in ** (-0.5)
+
+
+class InvariantMLP(torch.nn.Module):
+
+    def __init__(self, n_inputs: int, n_layers: int) -> None:
+        super().__init__()
+        # the last linear layer is applied outside of the MLP
+
+        # if there is more than one layer, expand the input dimension to 4 times its
+        # size and then reduce it back to the original size
+        layers = (
+            [Linear(n_inputs, n_inputs), torch.nn.SiLU()]
+            if n_layers == 1
+            else [Linear(n_inputs, 4 * n_inputs), torch.nn.SiLU()]
+            + [Linear(4 * n_inputs, 4 * n_inputs), torch.nn.SiLU()] * (n_layers - 2)
+            + [Linear(4 * n_inputs, n_inputs), torch.nn.SiLU()]
+        )
+        self.mlp = torch.nn.Sequential(*layers)
+
+    def forward(self, features: TensorMap) -> TensorMap:
+
+        # assume invariant
+        block = features.block({"o3_lambda": 0, "o3_sigma": 1})
+
+        output_values = self.mlp(block.values)
+        new_block = TensorBlock(
+            values=output_values,
+            samples=block.samples,
+            components=block.components,
+            properties=block.properties,
+        )
+
+        return TensorMap(
+            keys=Labels(
+                names=["o3_lambda", "o3_sigma"],
+                values=torch.tensor(
+                    [[0, 1]], dtype=torch.int32, device=new_block.values.device
+                ),
+            ),
+            blocks=[new_block],
+        )
+
+
+class EquivariantLinear(torch.nn.Module):
```

Comment:
can't you use the version in metatensor-learn?

Reply:
It needs the custom `Linear` that is defined above, which initializes the weights in a different way from pure torch